### PR TITLE
add test for checking if metrics are conformant according to promtool

### PR DESCRIPTION
**Pull request by coffeepac**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This is the optional tooling for CI as described in the metric overhaul [KEP](https://github.com/kubernetes/enhancements/blob/73a9f147245bee11d6590e9c6e09c6ee349bcac7/keps/sig-instrumentation/20181106-kubernetes-metrics-overhaul.md#graduation-criteria)

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
promtool rules are specified in code [here](https://github.com/prometheus/prometheus/blob/master/util/promlint/promlint.go) . We need to decide if we want to use all of these or if we are going to create a stripped down version.

We should also consider adding an exceptions list of all current metrics and fixing them as and when we are able.

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
[KEP]: https://github.com/kubernetes/enhancements/blob/73a9f147245bee11d6590e9c6e09c6ee349bcac7/keps/sig-instrumentation/20181106-kubernetes-metrics-overhaul.md
```

### CODE REVIEW

**Comment by coffeepac** [comment]
I expect we will discuss this at KubeCon if we have time.

**Comment by brancz** [comment]
I assume we violate a bunch of these lints. Do you know if we can extract a list of metrics that failed it? I assume that those would be candidates for a next round of cleanup.

**Comment by coffeepac** [comment]
we violate a bunch of these.  for now the build is failing because I'm not real good with bazel.

this line:  https://github.com/kubernetes/kubernetes/pull/85173/files#diff-b0b41100787a01b9a86aa3ad23454c7bR52 has a variable named 'problems'.  We can add custom filtering there for what we consider to be a failure.  this keeps us on the community standard but gives us some time to get up to speed.

**Comment by RainbowMango** [comment]
@brancz 
I tried to extract a list of metrics which not in line with the `promlint`.
There are about 22 metrics matched. More details you can refer to [here](https://github.com/RainbowMango/MetricLinter/tree/master/report).

> We can add custom filtering there for what we consider to be a failure. this keeps us on the community standard but gives us some time to get up to speed.

Basically, I agree with @coffeepac.  Furthermore, I want to improve or refactor the framework, now just a few metrics are covered by it. Anyway, I need to investigate more first.

**Comment by brancz** [comment]
I would say we need to get the tooling in place as soon as we can. We know we have current violations, so my suggestion would be to somehow allow and not fail on the current violations but on any new ones.

**Comment by RainbowMango** [comment]
Yeah, So, please @coffeepac continue your terrific work. And If you need help, just feel free to let me know.

**Comment by cblecker** [comment]
The bazel piece can be fixed by `hack/update-bazel.sh` and committing the changes.

File: test/e2e/instrumentation/monitoring/metric_conformance.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		framework.ExpectNoError(err)
+		grabber, err = metrics.NewMetricsGrabber(c, ec, true, true, true, true, true)
+		framework.ExpectNoError(err)
+	})
+
+	gin.It("all metrics from API server should conform to promtool.", func() {
+		gin.By("Connecting to /metrics endpoint")
+		apimetrics, err := grabber.GrabRawFromAPIServer()
+		framework.ExpectNoError(err)
+		gom.Expect(apimetrics).NotTo(gom.BeEmpty())
```
Comment (RainbowMango) [inline test/e2e/instrumentation/monitoring/metric_conformance.go]: `gom.Expect(foo).To(gomega.Equal(bar))` should be replaced by `framework.ExpectEqual(foo, bar)` which is more readable.  More details please refer to https://github.com/kubernetes/kubernetes/issues/79686.

Comment (RainbowMango) [reply test/e2e/instrumentation/monitoring/metric_conformance.go]: Same as below.

**Review by RainbowMango — COMMENTED** [review]
Just a nit.

File: test/e2e/instrumentation/monitoring/metric_conformance.go (hunk 1 above)
Comment (coffeepac) [reply test/e2e/instrumentation/monitoring/metric_conformance.go]: Thanks for the notes!  I plan on working on this soon.  Got a work related deadline of tomorrow that's been eating all my time.

**Comment by RainbowMango** [comment]
@brancz I want to introduce a wrapper of `promlint`, which can be placed in `component-base\metrics\testutil`. The benefits we can get from it:
- Ignore individual metrics for those current violations; 
- Lint items and behavior can be easily personalized; 
- Any exception will cause sig-instrumentation to be tagged for approvals; 
- Prevent other components directly reference Prometheus. 

How do you think?

**Comment by brancz** [comment]
@RainbowMango sounds like a great idea!

**Comment by RainbowMango** [comment]
OK. I will send a draft PR soon.

**Comment by ehashman** [comment]
Where are we with this and #86477?

**Comment by coffeepac** [comment]
@ehashman do we need both?  I thought @RainbowMango's PR covered the same ground.

**Comment by ehashman** [comment]
That's why I'm asking :)

**Comment by RainbowMango** [comment]
#86477 intended to provide common functions by which we can use in E2E like this PR.
#86477 is blocked now and I'm trying to provide a new solution.
So, I'd like to keep the two PR for a while, I think they can be done by 1.19.

**Comment by coffeepac** [comment]
@RainbowMango okay, I'll try and get back to this soon.

**Comment by RainbowMango** [comment]
@coffeepac You can continue now, and sorry for let this sit so long. It's quite a ride.

I've figured out some metrics violate `promlint` rules and listed in [an exception list](https://github.com/kubernetes/kubernetes/blob/d66bbd87283fd5219f84e89ccafb6cffe3396171/staging/src/k8s.io/component-base/metrics/testutil/promlint.go#L30). Since I grab them from a local-up cluster, maybe it does not contain all metrics, so if you find more, you can append them to the list(they will be fixed later) or just fix them in this PR.

**Comment by RainbowMango** [comment]
/remove-lifecycle stale
Kindly ping @coffeepac

**Comment by coffeepac** [comment]
@RainbowMango thank you for the kindly ping.  I remain constrained by covid-19 still.  If someone wants to take this over I would not object but this is still on my list (haven't closed the tab with the PR in over a year.  gosh.)

**Comment by ehashman** [comment]
@coffeepac I will only condone removing stale if you update this PR :joy:

**Comment by coffeepac** [comment]
@ehashman both fair and correct!  I won't remove stale again unless its with a commit.  I imagine my next rebase will be...  "fun".

**Comment by coffeepac** [comment]
this isn't going to work as currently constructed due to https://github.com/kubernetes/kubernetes/issues/89267, I need to refactor to use the Kubernetes layer.

File: test/e2e/instrumentation/monitoring/metric_conformance.go (hunk 1 above)
Comment (coffeepac) [reply test/e2e/instrumentation/monitoring/metric_conformance.go:49]: @RainbowMango  never forgot about this.  thanks covid.

I don't know how to replace this usage of `gom.Expect`.  I was unable to find a way to express 'explode if target string is empty' via the framework package.  the below `gom.Expect` was updated.

**Comment by dims** [comment]
/retest

@coffeepac looks like there are a bunch of comments from reviewers. Can you please take a look?

**Pull request closed without merging**